### ISSUE TITLE
 translate_el.sh: fail to chroot

### ISSUE THREAD

**Issue by helen-fornazier** [GoogleCloudPlatform/compute-image-tools#130]
For both Centos 7 and RHEL 7 I get the error below, I'll dig more to find a solution.
The exist point is when it tries to execute the following and fails:
  `chroot ${MNT} yum -y install google-compute-engine google-compute-engine-init google-config`
Other errors are also present

```
Debian GNU/Linux 9 inst-translator-translate-centos-7-translate-disk-stvzt ttyS0

inst-translator-translate-centos-7-translate-disk-stvzt login: Aug 22 20:52:46 localhost google_metadata_script_runner[705]: Copying gs://main-nucleus-128012-daisy-bkt/daisy-translate-centos-7-20170822-20:51:58-stvzt/sources/translate_el.sh...
Aug 22 20:52:46 localhost google_metadata_script_runner[705]: / [0 files][    0.0 B/  4.0 KiB]                                                #015/ [1 files][  4.0 KiB/  4.0 KiB]
Aug 22 20:52:46 localhost google_metadata_script_runner[705]: Operation completed over 1 objects/4.0 KiB.
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + URL=http://metadata/computeMetadata/v1/instance/attributes
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: ++ curl -f -H Metadata-Flavor:Google http://metadata/computeMetadata/v1/instance/attributes/el_release
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url:                                  Dload  Upload   Total   Spent    Left  Speed
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: #015  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0#015100     1  100     1    0     0     66      0 --:--:-- --:--:-- --:--:--    71
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + EL_RELEASE=7
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: ++ curl -f -H Metadata-Flavor:Google http://metadata/computeMetadata/v1/instance/attributes/install_gce_packages
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url:                                  Dload  Upload   Total   Spent    Left  Speed
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: #015  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0#015100     4  100     4    0 [   37.194660] SGI XFS with ACLs, security attributes, realtime, no debug enabled
    0    671      0 --:--:-- --:[   37.205338] XFS (sdb1): Mounting V4 Filesystem
--:-- --:--:--   800
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + INSTALL_GCE=true
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: ++ curl -f -H Metadata-Flavor:Google http://metadata/computeMetadata/v1/instance/attributes/use_rhel_gce_license
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url:                                  Dload  Upload   Total   Spent    Left  Speed
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: #015  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0#015100     5  100     5    0     0    835      0 --:--:-- --:--:-- --:--:--  1000
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + RHEL_LICENSE=false
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + MNT=/mnt/imported_disk
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + mkdir /mnt/imported_disk
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + '[' -b /dev/sdb1 ']'
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + echo 'Trying to mount /dev/sdb1'
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: Trying to mount /dev/sdb1
Aug 22 20:52:46 localhost startup-script: INFO startup-script-url: + mount /dev/sdb1 /mnt/imported_disk
Aug 22 20:52:47 localhost kernel: [   37.194660] SGI XFS with ACLs, security attributes, realtime, no debug enabled
Aug 22 20:52:47 localhost kernel: [   37.205338] XFS (sdb1): Mounting V4 Filesystem
[   38.631959] random: crng init done
Aug 22 20:52:48 localhost kernel: [   38.631959] random: crng init done
[   40.274745] XFS (sdb1): Ending clean mount
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + '[' 0 -ne 0 ']'
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + for f in proc sys dev run
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + mount -o bind /proc /mnt/imported_disk/proc
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: mount: mount point /mnt/imported_disk/proc does not exist
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + for f in proc sys dev run
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + mount -o bind /sys /mnt/imported_disk/sys
Aug 22 20:52:50 localhost kernel: [   40.274745] XFS (sdb1): Ending clean mount
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: mount: mount point /mnt/imported_disk/sys does not exist
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + for f in proc sys dev run
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + mount -o bind /dev /mnt/imported_disk/dev
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: mount: mount point /mnt/imported_disk/dev does not exist
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + for f in proc sys dev run
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + mount -o bind /run /mnt/imported_disk/run
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: mount: mount point /mnt/imported_disk/run does not exist
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + cp /etc/resolv.conf /mnt/imported_disk/etc/resolv.conf
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: cp: cannot create regular file '/mnt/imported_disk/etc/resolv.conf': No such file or directory
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + oot /mnt/imported_disk restorecon /etc/resolv.conf
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: /startup-N0dtKF/tmpnDsYM_: line 47: oot: command not found
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + [[ '' == \t\r\u\e ]]
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + [[ true == \t\r\u\e ]]
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + echo 'Installing GCE packages.'
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: Installing GCE packages.
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + cat
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: /startup-N0dtKF/tmpnDsYM_: line 64: /etc/yum.repos.d/google-cloud.repo: No such file or directory
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + [[ 7 == \7 ]]
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + cat
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: /startup-N0dtKF/tmpnDsYM_: line 76: /etc/yum.repos.d/google-cloud.repo: No such file or directory
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + chroot /mnt/imported_disk yum -y install google-cloud-sdk
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: chroot: failed to run command â€˜yumâ€™: No such file or directory
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + chroot /mnt/imported_disk yum -y install google-compute-engine google-compute-engine-init google-config
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: chroot: failed to run command â€˜yumâ€™: No such file or directory
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + '[' 127 -ne 0 ']'
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + echo 'TranslateFailed: GCE package install failed.'
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: TranslateFailed: GCE package install failed.
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: + exit 1
Aug 22 20:52:50 localhost startup-script: INFO startup-script-url: Return code 1.
Aug 22 20:52:50 localhost startup-script: INFO Finished running startup scripts.
Aug 22 20:52:50 localhost systemd[1]: Started Google Compute Engine Startup Scripts.
Aug 22 20:52:50 localhost systemd[1]: Reached target Multi-User System.
Aug 22 20:52:50 localhost systemd[1]: Reached target Graphical Interface.
Aug 22 20:52:50 localhost systemd[1]: Starting Update UTMP about System Runlevel Changes...
Aug 22 20:52:50 localhost systemd[1]: Started Update UTMP about System Runlevel Changes.
Aug 22 20:52:50 localhost systemd[1]: Startup finished in 1.170s (kernel) + 38.823s (userspace) = 39.993s.


```

**Comment by zmarano** [comment]
Is this an LVM disk or a disk with the boot loader on a separate partition from the root file system? If so, that would be why. Right now everything just fails miserably in those cases. We actually need to do a bunch more work [here](https://github.com/GoogleCloudPlatform/compute-image-tools/blob/master/daisy_workflows/image_import/enterprise_linux/translate_el.sh#L28) to sort out if we have mounted a valid root partition and where the boot loader is- if its not there maybe its on another partition etc. If you have any ideas how to sanely do that we're open :) I haven't come up with anything that makes sense yet.

**Comment by helen-fornazier** [comment]
I see, yes, I have /boot in a different partition indeed.
The rootfs in a linux system needs to have at least the file /sbin/init (imposed by the kernel), so it shouldn't be too difficult to identify which partition rootfs is by looping through all the partitions, mounting them and searching for this file, it should also be possible to mount a LVM disk and check its content (I'll do a better research on this).
For the boot partition, the script already assumes that grub is being used, so we could do the same and look for the grub.cfg file.
Another solution is to ask the user to run a helper/hint tool in the machine to analyze the system and upload the collected information to a gs bucket, but this requires one more interaction from the user (not good for UX).
We could also parse /etc/fstab, but if it uses labels instead of UUIDs for the disks, then it is not reliable, but we could use it as a hint.

**Comment by zmarano** [comment]
I was hoping to come up with some common library functions to do this- because all of the Linux imports need to figure out the same thing. I was looking at https://pypi.python.org/pypi/lvm2py/ as a possibility for LVM but it wasn't very helpful.

**Comment by helen-fornazier** [comment]
The only usage I see about using a common library is to retrieve the FS uuid or label, the partitions would still need to be mounted in the right location to allow chroot. Let me know if you were thinking in something else.
I was thinking in something like this https://gist.github.com/helen-fornazier/4cc431d79c523d8cfb4131e8d19c833f
i.e, to probe all the partitions, and then mount the rootfs and the boot in the right place to allow chroot.
This still need to be polished and discussed, but this would solve the LVM problem and also the /boot in a separated partition, the limitation is still if the system have other weird partitions, but this covers most of the cases.
What do you think?

**Comment by helen-fornazier** [comment]
Or we could use libguestfs (there are python and go libraries for that), with it we can find the rootfs with the function inspect_os(), check the mount points with inspect_get_mountpoints() and also mount the disk. This seems more portable and reliable.

http://libguestfs.org/
http://libguestfs.org/guestfs-golang.3.html
http://libguestfs.org/guestfs-python.3.html

**Comment by helen-fornazier** [comment]
We could also use guestfs to manage this.
I was able to mount the entire disk using
$ guestmount --pid-file guestmount.pid -a /dev/sdb -i /mnt
There are libraries in several languages, python and go included.
This tool uses the help of qemu to access the filesystem, it should be more reliable then the previous suggested approach.
There are some other ways to access the filesystem and to customize it without mounting it in the host using this tool, I am digging a bit more into this.

http://libguestfs.org/

**Comment by zmarano** [comment]
That looks promising. I never heard of guestfs- lets see if that will work for the common cases.

**Comment by zmarano** [comment]
I played around with the CLI tools a bit. virt-filesystems is sort of useful in that it can identify all the LVM volumes. It looks like you still need to mount everything and figure out where the boot loader lives. It is also a bit wonky in terms of displaying block devices with filesystems. For example, from a CentOS disk I have around.

```
~# virt-filesystems --filesystems -a /dev/sdb
/dev/sda1
/dev/centos/root
```

/dev/sda1 contains the boot loader (although in this case, its actually /dev/sdb1)
/dev/centos/root contains the root filesystem on an LVM volume

**Comment by helen-fornazier** [comment]
I tested with
$ guestmount -a /dev/sdb -i /mnt
I was able to complete the translate script with success, but I encountered other problems along the way:
* For Debian with LVM: It mounted the disk, ran the chroot commands and finished successfully, but the disk didn't boot, I think this is another problem not directly related to guestmount, but to the boot configs, I am still investigating this.

* For Centos 7: The script has other problems (wrong prefix and paths), and the boot config.
I was able to finish the translate script and boot the machine partially because the major problem is when it has an initramfs which needs to be regenerated with the tool dracut, the problem is that dracut checks which kernel it runs so it can look for the modules at /lib/modules/<kern-version>, but as it is running through chroot, it get the wrong kernel information (basically it executes uname -r). I could probe which kernel version it uses and execute the command:
$ dracut -f --kern-ver 3.10.327.el # For example
But this doesn't seem ideal
So I am checking the virt-customize tool, it uses qemu to access the system, and it allows me to execute a script inside the vm machine to install packages and do other changes. I am still investigating if it solves the entire problem but seems promising.
So the idea would be, instead of mounting the disk and performing changes with chroot, we could use virt-customize directly to make all the changes, as it is in a isolated qemu environment, it should be more reliable.
For example, it is possible to execute a command with:
$ virt-customize -v -a /dev/sdb --run-command "ls -1 /"
It shouldn't matter if the disk has multiple partitions with or without LVM. I am doing more tests.
http://libguestfs.org/virt-customize.1.html#log-file

In the python-guestfs library, there is also the possibility to execute shell script inside the VM to customize it. So I was thinking we could write a script in bash, python or golang or any language you prefer that has a guestfs library (which language do you prefer?) and use virt-customize to mount and customize the image, if I understand correct (I am still doing more tests), this should deal with the LVM problems, several partition problems and also the `uname -r` problem.
What do you think?

**Comment by zmarano** [comment]
Oh yeah that worked!

Debian translations- are you sure its not booting or just not outputting anything to the serial console (which is a grub kernel cmd line option).

For CentOS, yes I found several bugs in the current script (will make a PR shortly to address those). This is the first time I have been able to test it on an actual imported disk. Will have to look into this dracut problem a bit more though.

I will have to think about the idea of using qemu- we don't currently have the ability of running a KVM VM inside of a GCE VM.

Language wise- I don't have a strong preference but given that we try to stay close to what our users would know, bash or python are preferable here. I did have intention to make common python libraries for all of these things and share them between the various translations.

**Comment by zmarano** [comment]
FYI: https://github.com/GoogleCloudPlatform/compute-image-tools/pull/143

**Comment by helen-fornazier** [comment]
The problem of using guestmount, is that grub2-mkconfig generates the wrong rootfs parameter, so the system doesn't boot and only the dracut shell is loaded. Using virt-customize for this solves the problem, but it is not that easy to pass a bash script to virt-customize.
So I wrote a script in python that seems to solve most of our problems, the only ugly thing is that I mixed bash with python for now to be able to install the python libraries before executing the python script (but this is temporary, only for the proof of concept)

@zmarano please check: https://github.com/helen-fornazier/compute-image-tools/blob/translate_el/daisy_workflows/image_import/enterprise_linux/translate_el.sh

To not mix bash with python in the same file, I was thinking in adding the python in the `Source` list in translate_el.wf.json (a translate_el.py besides the translate_el.sh), but I am not exactly sure how to retrieve the translate_el.py from the translate_el.sh script inside the staging VM, do you have any recommendation regarding this?

**Comment by helen-fornazier** [comment]
The problem of using guestmount is that grub2-mkconfig generates the wrong rootfs parameter, so the system doesn't boot and only the dracut shell is loaded. Using virt-customize for this solves the problem, but it is not that easy to pass a bash script to virt-customize.
So I wrote a script in python that seems to solve most of our problems and seems to be working fine.
@zmarano please see #148

**Comment by zmarano** [comment]
Resolved by #148
